cluster/gce/util.sh

```diff
@@ -531,16 +533,87 @@ function build-node-labels {
   echo $node_labels
 }
 
+# yaml-map-string-stringarray converts the encoded structure to yaml format, and echoes the result
+# under the provided name. If the encoded structure is empty, echoes nothing.
+# 1: name to be output in yaml
+# 2: encoded map-string-string (which may contain duplicate keys - resulting in map-string-stringarray)
+# 3: file to write to
+# 4: key-value separator (defaults to ':')
+# 5: item separator (defaults to ',')
+function yaml-map-string-stringarray {
+  declare -r name=$1
+  declare -r encoded=$2
+  declare -r file=$3
+  declare -r kv_sep=${4:-':'}
+  declare -r item_sep=${5:-','}
+
+  declare -a pairs # indexed array
+  declare -A map # associative array
+  IFS="${item_sep}" read -ra pairs <<<"${encoded}" # split on item_sep
+  for pair in "${pairs[@]}"; do
+    declare key
+    declare value
+    IFS="${kv_sep}" read -r key value <<<"${pair}" # split on kv_sep
+    map[$key]="${map[$key]+${map[$key]}${item_sep}}${value}" # append values from duplicate keys
+  done
+  # only output if there is a non-empty map
+  if [[ ${#map[@]} -gt 0 ]]; then
+    echo "${name}:"
+    for k in "${!map[@]}"; do
+      echo "  ${k}"
+      declare -a values
+      IFS="${item_sep}" read -ra values <<<"${map[$k]}"
+      for v in "${values[@]}"; do
+        echo "    - $(yaml-quote "${v}")"
```

Comment:
nitpick: I'd use an assignment to avoid the nested quotes:

```shell
  declare quoted_value
  quoted_value=$(yaml-quote "${v}")
  echo "    - ${quoted_value}"
```

Note that this doesn't work!

```shell
  declare quoted_value=$(yaml-quote "${v}")
```

Because if `yaml-quote` fails (returns non-zero) then that gets ignored since neither errexit or pipefail is triggered... :disappointed:

Reply:
done